HIP/hip-1056.md

```diff
@@ -0,0 +1,1933 @@
+creation logic allows for the consensus node or block validator to calculate the input and output tree merkle hashes as
+items are received in a streaming way.
+
+In summary a consensus node can calculate the merkle hashes of the transaction inputs and outputs in real time and need only calculate the state merkle root at the end of a block.
+
+**Example Pseudo Code for computing streaming merkle tree root hash**
+
+```jsx
+elements = [ ... ] // array of the leaves to be streamed and hashed. Items may be transaction inputs, transaction outputs or state
+hashList = [] // a new empty list of hashes
+
+// loop over each element received and create a list of all the root hashes of the leaves of the tree 
+for (int i=0, i<elements.size(), i++) {
+    write elements[i] // output elemen to stream
+    e = hash(elements[i]) // SHA 384 item data
+    hashList.add(e) // add hash to end of list
+    
+    // for every odd located element remove and hash the last 2 hashes (representing the hash leaves) and shift right per loop to pick up the hash root
+    for (int n=i, (n & 1 == 1), n>>=1) {
+        y = hashList.getAndRemove(hashList.size() – 1); // remove last element of hashList
+        x = hashList.getAndRemove(hashList.size() – 1); // remove last element of hashList
+        hashList.add(hash(x, y)); // add the new hash root of the two hashes
+    }    
+}
+
+merkleRootHash = hashList.get(hashList.size() – 1)
+
+for (int i=hashList.size() – 2, i>=0, i--) {
+    merkleRootHash = hash(hashList.get(i), merkleRootHash)
+}
+
+write merkleRootHash; // output
+```
+
+
+#### RecordFileItem
+
+A `RecordFileItem` block stream item is added to act as a wrapper around the contents of a legacy block. This allows
+block nodes and other handlers of the block stream to have single standardized APIs and handling code for blocks before
+and after the switch to block streams. It is not possible to change the format because it is signed by the consensus
+nodes in the original format. As a result of this a reader will still have to understand record file formats version 2,
+3, 5 and 6 that existed prior to block streams. Each block of record style will have just one item: a `RecordFileItem`.
+
+```protobuf
+/**
+ * A Block Item for record files.
+ */
+message RecordFileItem {
+  /**
+  * The block number of this block.
+  */
+  uint64 number = 1;
+
+  /**
+  * The consensus time the record file was produced for.<br/>
+  * This comes from the record file name.
+  */
+  proto.Timestamp creation_time = 2;
+
+  /**
+  * The contents of a record file.<br/>
+  * The first 4 bytes are a 32bit int little endian version number.
+  * The versions that existed are 2,3,5 and 6.
+  */
+  bytes record_file_contents = 3;
+
+  /**
+  * The contents of sidecar files for this block.<br/>
+  * Each block can have zero or more sidecar files.
+  */
+  repeated bytes sidecar_file_contents = 4;
```

Comment:
It would require parsing the sidecar file data when building these from the existing file data, and that's already a _very_ long and difficult process.
@jasperpotts was working on that last I heard, so you could ask if he's willing to make that change.

Reply:
100% lets make it SidecarFile much nicer. Easy enough to fix importer.

Reply:
Updated